### PR TITLE
FEAT: Add Badam optimizer

### LINKED ISSUE #30308: Badam support
### Feature request

https://arxiv.org/pdf/2404.02827.pdf

Memory efficient optimizer like galore but with less hyperparameters 

### Motivation

https://github.com/hiyouga/LLaMA-Factory/pull/3287

LlamaFactory supports it, but having it in transformers is much more convenient

### Your contribution

-

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes: https://github.com/huggingface/transformers/issues/30308

This PR adds Badam optimizer to transformers Trainer API 

## TODOs:

- [ ] Add ratio optimizer
- [x] Add layer optimizer
- [ ] Add docs

cc @amyeroberts @muellerzr

### CODE REVIEW

**Review by muellerzr — APPROVED** [review]
Nice job :) Very straightforward.

File: src/transformers/training_args.py [hunk 1]
```diff
@@ -1620,7 +1620,15 @@ def __post_init__(self):
...
                 raise ValueError("lr_scheduler_type reduce_lr_on_plateau requires torch>=0.2.0")
 
+        use_badam = "badam_" in self.optim
+        if use_badam:
+            self.optim = self.optim.replace("badam_", "")
+
         self.optim = OptimizerNames(self.optim)
+
+        if use_badam:
+            self.optim = "badam_" + self.optim
```
Comment (amyeroberts) [inline src/transformers/training_args.py:1630]: Having to remove and then add this back seems both convoluted and an indication that there's something funny about the implementation. Specifically, it seems like something which should be handled on the `OptimizerNames` side

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1079,6 +1080,16 @@ def create_optimizer(self):
...
 
+            if badam_kwargs is not None:
+                from badam import BlockOptimizer
+
+                self.optimizer = BlockOptimizer(
+                    base_optimizer=self.optimizer,
+                    named_parameters_list=list(opt_model.named_parameters()),
+                    block_prefix_list=None,
+                    **badam_kwargs,
+                )
```
Comment (amyeroberts) [inline src/transformers/trainer.py:1091]: This is another indication of a peculiar pattern - why set `self.optimizer` and then overwrite here? Instead, the correct optimizer class and kwargs should be handled in `Trainer.get_optimizer_cls_and_kwargs(self.args, opt_model)`

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -1126,13 +1137,20 @@ def get_optimizer_cls_and_kwargs(
...
+            args.optim = args.optim.replace("badam_", "")
 
         # parse args.optim_args
         optim_args = {}
+        badam_optim_args = {}
         if args.optim_args:
             for mapping in args.optim_args.replace(" ", "").split(","):
                 key, value = mapping.split("=")
-                optim_args[key] = value
+                if "badam_" in key and use_badam:
```
Comment (amyeroberts) [inline src/transformers/trainer.py:1150]: Is there ever a case when `use_badam` is `False` and `badam_` is in the key?

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -1126,13 +1137,20 @@ def get_optimizer_cls_and_kwargs(
...
+                    badam_optim_args[key.replace("badam_", "")] = value
+                else:
+                    optim_args[key] = value
```
Comment (amyeroberts) [inline src/transformers/trainer.py:1153]: We shouldn't need to do this string manipulation and have these parallel optim_args. This doesn't scale well if we want to add other optimizers which accept some of the previous optim args

File: src/transformers/trainer.py (hunk 3 above)
Comment (amyeroberts) [inline src/transformers/trainer.py:1150]: nit - check the bool flag first in the and check, it'll be faster 

```suggestion
                if use_badam and "badam_" in key:
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for the work adding this! 

At the moment, I think there's some work needed to move the handling logic into different methods to prevent leakiness i.e. not all of these methods need to know about badam. 

p.s. I can't read badam and not think about [this Kylie banger](https://youtu.be/p6Cnazi_Fi0?si=0cxt3rZh35HQF2Ya)

**Comment by Ledzy** [comment]
Thank you for your efforts in integrating our work! 

BAdam supports model parallel offered by Deepspeed ZeRO-3 now. Integrating the latest feature only requires 
* set `ds_zero3_enabled` when initializing the `BlockOptimizer`
* add a customized `BAdamCallback` object

Please see our implementation [here](https://github.com/Ledzy/BAdam). It enable doing full parameter finetune on a 70B model with only 8 RTX-3090 GPUs or 3 A100 GPUs. In terms of time and computational efficiency, BAdam saves half backward FLOPs in average and thereby requires less training time compared with some existing PEFT approaches, e.g. LoRA.

The current ratio-based implementation is not well optimized and requires computing the full gradient of the model for each backward (it sparsifies the gradient on the fly). We left the implementation-level optimization as a future work, and suggest not including the ratio-based implementation for now.

Should you encounter any issue when integrating BAdam, please feel free to let us know. We are more than happy to assist in the integration, and are willing to change the source code if needed.
